site/content/protocols/messagepickup/4.0/readme.md

```diff
@@ -0,0 +1,388 @@
+
+Delivered messages **MUST NOT** be deleted until delivery is acknowledged by a `messages-received` message.
+
+### Message Delivery
+Batch of messages delivered to the `recipient` as attachments.
+
+Message Type URI: `https://didcomm.org/message-pickup/4.0/delivery`
+
+DIDComm v1 example:
+```json
+{
+    "@id": "123456780",
+    "@type": "https://didcomm.org/message-pickup/4.0/delivery",
+    "~thread": {
+        "thid": "<message id of delivery-request message>"
+    },
+    "recipient_did": "<did:key for messages>",
+    "~attach": [{
+        "@id": "<id of message>",
+        "data": {
+            "base64": "<message>"
+        }
+    }]
+}
+```
+
+DIDComm v2 example:
+```json
+{
+    "id": "123456780",
+    "thid": "<message id of delivery-request message>",
+    "type": "https://didcomm.org/message-pickup/4.0/delivery",
+    "body": {
+        "recipient_did": "<did for messages>"
+    },
+    "attachments": [{
+        "id": "<id of message>",
+        "data": {
+            "base64": "<message>"
+        }
+    }]
+}
+```
+
+Messages delivered from the queue **MUST** be delivered in a batch delivery message as attachments, with a batch size specified by the `limit` provided in the `delivery-request` message.
+The `id` of each attachment is used to confirm receipt.
+The `id` is an opaque value, and the recipient **SHOULD** not deduce any information from it, except that it is unique to the mediator. The recipient can use the `id`s in the `message_id_list` field of `messages-received`.
+
+If there are no messages in the queue for the `delivery-request`, the `delivery` message **MUST** contain an empty array of attachments.
+
+The ONLY valid type of attachment for this message is a DIDComm v1 or v2 Message in encrypted form.
+
+`thid` -- an optional field if the delivery message is in response to a singular `delivery-request` messsage.
+
+The `recipient_did` attribute is only included when responding to a `delivery-request` message that indicates a `recipient_did`.
+
+### Messages Received
+After receiving messages, the `recipient` **MUST** send an acknowledge message indiciating which messages are safe to clear from the queue.
+
+Message Type URI: `https://didcomm.org/message-pickup/4.0/messages-received`
+
+DIDComm v1 example:
+```json
+{
+    "@id": "123456780",
+    "@type": "https://didcomm.org/message-pickup/4.0/messages-received",
+    "message_id_list": ["123","456"]
+}
+```
+
+DIDComm v2 example:
+```json
+{
+    "id": "123456780",
+    "type": "https://didcomm.org/message-pickup/4.0/messages-received",
+    "body": {
+        "message_id_list": ["123","456"]
+    }
+}
+```
+
+`message_id_list` is a list of `ids` of each message received. The `id` of each message is present in the attachment descriptor of each attached message of a delivery message.
+
+Upon receipt of this message, the `mediator` knows which messages have been received, and can remove them from the collection of queued messages with confidence.
```

Comment:
I noticed in this section that it does not say anything about an updated `status` message sent as response for this message. This was explicitly mentioned in previous versions, so I'm wondering if this update intends to reflect that no further response is expected, or if it's just to not be too redundant (in Basic Walkthrough it is already stated it is a response)

Reply:
Good catch -- that should definitely go in the changelog

I believe this is motivated by the same complexity / ambiguity as to why we removed a `status` message as a followup to a `delivery-request` when no messages are queued for delivery, as it is unclear as to if it should be filtered by `recipient_did` or not (and actually can't be).
For example, in the following flow: `status-request` (restricted by `recipient_did`) -> `status` -> `delivery` -> `messages-received` -> `status` (NOT restricted by `recipient_did`, as there's no way to know that we were doing `messages-received` in regards to a filtered flow)
This may give the incorrect view that a given `recipient_did` has additional messages awaiting for pickup, but in reality they are for a different `recipient_did`. 

If I'm missing something here, please say so

Reply:
I've somewhat fixed the walkthrough section (cause yeah that's confusing, nice catch).

>  if I have 10 messages and my limit is 10, I'll need to do a status-request -> status -> delivery-request -> delivery -> messages-received followed by a status-request -> status. 

That's only if you assume that you could have received a new message in the 200ms or however long it takes to receive the 10 messages (because we were informed by the first `status` message that we had 10 at that point in time). You in fact could receive a new message right after sending the second status-request and think you have no messages queued. I think it's an inherit problem with polling strategies.